cpp/src/scalar/scalar_factories.cpp

```diff
@@ -62,4 +62,18 @@ std::unique_ptr<scalar> make_timestamp_scalar(
                                        stream, mr);
 }
 
+namespace {
+struct default_scalar_functor {
+  template <typename T>
+  auto operator()() {
+    using ScalarType = experimental::scalar_type_t<T>;
+    return std::unique_ptr<scalar>(new ScalarType);
+  }
+};
+}  // namespace
+
+std::unique_ptr<scalar> make_default_constructed_scalar(data_type type) {
+  return experimental::type_dispatcher(type, default_scalar_functor{});
```

Comment:
I've been trying out this patch and ran into some issues with reductions on empty columns.  I tracked it down to this code.

Calling the default constructor leaves the scalar type set to `EMPTY`.  That loses the specified type and results in a scalar that cannot be type-dispatched.  I suggest adding a unit test for this new scalar constructor that's similar to the `FactoryDefault` tests in factories_test.cpp.

Reply:
correct fix should be to update default constructed scalar to have respective `type` right? @jrhemstad @jlowe

Reply:
Thanks. Neat catch. I updated the default constructors.